test/unit/org/apache/cassandra/db/streaming/CassandraEntireSSTableStreamWriterTest.java

```diff
@@ -105,69 +124,144 @@ public static void defineSchemaAndPrepareSSTable()
+
+            return null;
+        });
+    }
+
+    /**
+     * Entire-sstable-streaming receiver will throw checksum validation failure because concurrent index summary
+     * redistribution task will modify index summary causing actual transfered index summary size to be different from
+     * the one in {@link ComponentManifest}
+     */
+    @Test
+    public void testBlockReadingAndWritingOverWireWithIndexSummaryRedistribution() throws Exception
+    {
+        testBlockReadingAndWritingOverWire(() -> {
+
+            long nonRedistributingOffHeapSize = 0;
+            long memoryPoolBytes = 1024 * 1024;
+
+            // rewrite index summary file with new min/max index interval
+            TableMetadata origin = store.metadata();
+            MigrationManager.announceTableUpdate(origin.unbuild().minIndexInterval(1).maxIndexInterval(2).build(), true);
+
+            try (LifecycleTransaction txn = store.getTracker().tryModify(Collections.singletonList(sstable), OperationType.INDEX_SUMMARY))
+            {
+                latch.countDown();
+                IndexSummaryManager.redistributeSummaries(new IndexSummaryRedistribution(ImmutableMap.of(store.metadata().id, txn),
+                                                                                         nonRedistributingOffHeapSize,
+                                                                                         memoryPoolBytes));
+            }
 
+            // reset min/max index interval
+            MigrationManager.announceTableUpdate(origin, true);
+            return null;
+        });
+    }
+
+    public void testBlockReadingAndWritingOverWire(Callable<?> componentMutation) throws Exception
+    {
         // This is needed as Netty releases the ByteBuffers as soon as the channel is flushed
         ByteBuf serializedFile = Unpooled.buffer(8192);
-        EmbeddedChannel channel = createMockNettyChannel(serializedFile);
-        AsyncStreamingOutputPlus out = new AsyncStreamingOutputPlus(channel);
+        InetAddressAndPort peer = FBUtilities.getBroadcastAddressAndPort();
+        StreamSession session = setupStreamingSessionForTest();
+        ComponentManifest manifest = CassandraOutgoingFile.getComponentManifest(sstable, true);
+        CassandraStreamHeader header = header(manifest);
 
-        writer.write(out);
+        sstable.runWithReadLock(ignored -> {
+            CassandraEntireSSTableStreamWriter writer = new CassandraEntireSSTableStreamWriter(sstable, session, manifest);
+
+            // start concurrent sstable component mutation
+            Future<?> mutation = CompletableFuture.completedFuture(null);
+            if (componentMutation != null)
+            {
+                mutation = ScheduledExecutors.optionalTasks.submit(componentMutation);
+                // wait for concurrent task to start
+                waitForLatch(10, TimeUnit.SECONDS);
```

Comment:
@jasonstack Out of curiosity, I disabled all the read/write lock acquisition in this patch, and the tests here all seem to pass pretty consistently still. I think what we might need to do here is take the concept here, and most of the code, and test w/ two threads, one streaming at the `CassandraOutgoingFile#write()` level, where our read locking actually occurs, and the other guaranteed to make a mutation to the stats or index summary between initial manifest creation and streaming. (i.e. Something that would clearly fail before this patch.)

I think the other thing that made this more apparent to me after looking a second time was the fact that we actually call `runWithReadLock()` explicitly in `testBlockReadingAndWritingOverWire()`. We should be testing the locking mechanisms in a more black box fashion, not replacing part of that mechanism in the test.

WDYT?

Reply:
agreed, we need another test to verify the lock. These tests in `CassandraEntireSSTableStreamWriter` only verify hard links are working as expected (`hard links` are not impacted by replace/deletion).

Reply:
agreed, we need some tests for locks. `CassandraEntireSSTableStreamWriter` only verifies hard links are working as expected, ie. `not affected by file deletion or replacement`